torch/distributed/optim/zero_redundancy_optimizer.py

```diff
@@ -147,23 +149,24 @@ def __init__(
         parameters_as_bucket_view: bool = False,
         **default: Any,
     ):
+        # Perform type and assumption checks on the input parameters
+        self._verify_and_init_params(params)
+        self._verify_same_param_device()
+        self._verify_same_dense_param_type()
+        self._device = self._all_params[0].device
+
         # Hold all the model params in the root .param_groups
         # NOTE: the default constructor uses `add_param_group` which is partially overloaded here
         # we introduce the `initialized` flag for be able to dissociate the behaviour of
         # `add_param_group` in between super() and ZeroRedundancyOptimizer
         self.initialized = False
-        super().__init__(params, default)
+        super().__init__(self._all_params, default)
 
-        # Partition information. lazy evaluation, computed if requested
-        self._per_device_params_cache: "OrderedDict[torch.device, List[List[Parameter]]]" = (
-            OrderedDict()
-        )  # device, rank, params
+        # Partition information (evaluated lazily)
         self._param_rank_cache: Dict[torch.Tensor, int] = {}
         self._param_to_index_cache: Dict[int, int] = {}
         self._partition_parameters_cache: List[List[Dict]] = []
-        self._index_to_param_cache: Dict[int, torch.Tensor] = {}
-        self._all_params = params
-        self._reference_is_trainable_mask = list(map(_is_trainable, self._all_params))
+        self._index_to_param_cache: List[torch.Tensor] = []
```

Comment:
`self._index_to_param` is used only in `load_state_dict()`, which loads an input `state_dict`, consisting of `"state"` and `"param_groups"`. The mapping given by `self._index_to_param` is used to identify a parameter from each key in the `"state"` part of `state_dict` so that (1) that parameter's state in the rank's sharded optimizer state can be updated if the rank is responsible for the parameter and (2) the entry in `state_dict["state"]` can be set to `None` if the rank is not responsible. 

I think here we are peeking into the `state_dict` abstraction and assuming that the `state_dict` packs the parameters in the linear order matching the construction of `self._index_to_params`. This seems to be needed to filter the input `state_dict` to only have non-`None` values for the given rank's parameters before calling `super().load_state_dict()`. As such, having this data structure for decoding `state_dict`s seems important even if a user never calls `add_param_group()`.

I may be not seeing the full picture though. Feel free to let me know.